modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2413,7 +2413,20 @@ def binary_op(self, op, right_frame, join_type="outer"):
             1, left_parts, lambda l, r: op(l, r), right_parts
         )
         new_columns = self.columns.join(right_frame.columns, how=join_type)
-        return self.__constructor__(new_frame, joined_index, new_columns, None, None)
+        return self.__constructor__(
+            partitions=new_frame,
+            index=joined_index,
+            columns=new_columns,
+            # In cases where the result of the operation is empty, the new
+            # frame can't deduce the dtypes from its partitions as usual,
+            # because it has no partitions. Instead we have to compute the
+            # dtypes by applying the operation to empty dataframes and passing
+            # the resulting dtypes to the new frame.
+            dtypes=op(
+                pandas.DataFrame(index=self.index, columns=self.columns),
+                pandas.DataFrame(index=right_frame.index, columns=right_frame.columns),
+            ).dtypes,
```

Comment:
Besides the fact that we're executing UDF `op` twice (which could be really expensive), this is still a very unreliable construction. The output type isn't based on the source frames types at all, it always gives the same type in the result here.
<details><summary>Example</summary>

```python
# Frames holding integer columns
>>> df1_i["a"].dtypes, df2_i["a"].dtypes
(dtype('int64'), dtype('int64'))
>>> (df1_i + df2_i).dtypes  # int + int = int
a    int64
dtype: object
# Frames holding strings:
>>> df1_s["a"].dtypes, df2_s["a"].dtypes
(dtype('O'), dtype('O'))
>>> (df1_s + df2_s).dtypes  # string + string = string
a    object
dtype: object
# trying to infer dtypes with the present logic:
# int + int = object -> wrong
>>> (pandas.DataFrame(index=df1_i.index, columns=df1_i.columns) + pandas.DataFrame(index=df2_i.index, columns=df2_i.columns)).dtypes
a    object
dtype: object
>>> (pandas.DataFrame(index=df1_s.index, columns=df1_s.columns) + pandas.DataFrame(index=df2_s.index, columns=df2_s.columns)).dtypes
a    object
dtype: object
```

</details>

As far as I know, we don't have support for empty frames at all, that's why handling empty cases are very tricky. They should probably go as default to pandas or even be ignored if they're not critical.

Reply:
@dchigarev you make a good point, however it is not so simple.

A common pattern we see is creating an empty dataframe and using `append` to add to it. There are lots of other usecases for empty dataframes even though I do not like them. Ultimately, we need a better way to handle them (maybe a new class?) but we can't just hand everything off to pandas because we don't get the dataframe back.